packages/@aws-cdk/core/lib/permissions-boundary.ts

```diff
@@ -0,0 +1,161 @@
+ * Options for binding a Permissions Boundary to a construct scope
+ */
+export interface PermissionsBoundaryBindOptions {}
+
+/**
+ * Apply a permissions boundary to all IAM Roles and Users
+ * within a specific scope
+ */
+export abstract class PermissionsBoundary {
+  /**
+   * The default bootstrap permission boundary
+   * {@link DefaultStackSynthesizer.DEFAULT_PERMISSIONS_BOUNDARY_NAME} will be applied
+   * to all IAM Roles and Users within the given scope
+   *
+   * @example
+   * declare const app: App;
+   * new Stage(app, 'ProdStage', {
+   *   permissionsBoundary: PermissionsBoundary.default(),
+   * });
+   */
+  public static default(): PermissionsBoundary {
+    return new PermissionsBoundaryManager(DefaultStackSynthesizer.DEFAULT_PERMISSIONS_BOUNDARY_NAME);
+  }
+
+  /**
+   * Apply a permissions boundary with the given name to all IAM Roles
+   * and Users created within a scope. The name can include the '${Qualifier}' string
+   * which will be replaced by the stack qualifier.
+   *
+   * @param name the name of the permissions boundary policy
+   *
+   * @example
+   * declare const app: App;
+   * new Stage(app, 'ProdStage', {
+   *   permissionsBoundary: PermissionsBoundary.fromName('my-custom-permissions-boundary'),
+   * });
+   */
+  public static fromName(name: string): PermissionsBoundary {
+    return new PermissionsBoundaryManager(name);
+  }
+
+  /**
+   * Apply a permissions boundary with the given ARN to all IAM Roles
+   * and Users created within a scope. The ARN can include the '${Qualifier}' string
+   * which will be replaced by the stack qualifier.
+   h
+   * @param arn the ARN of the permissions boundary policy
+   *
+   * @example
+   * declare const app: App;
+   * new Stage(app, 'ProdStage', {
+   *   permissionsBoundary: PermissionsBoundary.fromArn('arn:aws:iam::${AWS::AccountId}:policy/my-custom-permissions-boundary'),
+   * });
+   */
+  public static fromArn(arn: string): PermissionsBoundary {
+    let name;
+    if (!Token.isUnresolved(arn)) {
+      name = Arn.parse(arn);
+    }
+    return new PermissionsBoundaryManager(name?.resourceName, arn);
+  }
+
+  /**
+   * Apply the permission boundary to the given scope
+   *
+   * Different permission boundaries can be applied to different scopes
+   * and the most specific will be applied.
+   */
+  public abstract bind(scope: Construct, options?: PermissionsBoundaryBindOptions): void;
+}
+
+/**
+ * Apply a permissions boundary to all IAM Roles and Users
+ * within a specific scope
+ *
+ * A permissions boundary is typically applied at the `Stage` scope.
+ * This allows setting different permissions boundaries per Stage. For
+ * example, you may _not_ apply a boundary to the `Dev` stage which deploys
+ * to a personal dev account, but you _do_ apply the default boundary to the
+ * `Prod` stage.
+ *
+ * It is possible to apply different permissions boundaries to different scopes
+ * within your app. In this case the most specifically applied one wins
```

Comment:
There is really no way to check that since it's just a string of a policy arn (or name). Really I don't see anyone specifying a permissions boundary at anything more specific than a stack since 1 stack == 1 bootstrap (but who knows 🤷‍♂️)

I could make the `bind` method `@internal` I try to limit the implementation to the `permissionsBoundary` props in Stack/Stage. Users could always use an "escape hatch" by manually setting the context if they wanted to.

Reply:
I updated this to be internal and we only allow setting the permission boundary on the app, stack, or stage.